docs/library/bluetooth.rst

```diff
@@ -0,0 +1,306 @@
+characteristics, which each have a value. Characteristics can also contain
+descriptors, which themselves have values.
+
+These values are stored locally and can be read from or written to by a remote
+central device. Additionally, a peripheral can "notify" its value to a connected
+central via its connection handle.
+
+.. method:: BLE.gatts_register_services(services_definition)
+
+    Configures the peripheral with the specified services, replacing any
+    existing services.
+
+    ``services_definition`` is a list of *services*, where each *service* is a
+    two-element tuple containing a UUID and a list of *characteristics*.
+
+    Each *characteristic* is a two-or-three-element tuple containing a UUID, a
+    *flags* value, and optionally a list of *descriptors*.
+
+    Each *descriptor* is a two-element tuple containing a UUID and a flags value.
+
+    The *flags* are a bitwise-OR combination of the ``FLAGS_READ``,
+    ``FLAGS_WRITE`` and ``FLAGS_NOTIFY`` values defined below.
+
+    The return value is a list (one element per service) of tuples (each element
+    is a value handle). Characteristics and descriptor handles are flattened
+    into the same tuple, in the order that they are defined.
+
+    The following example that registers two services (Heart Rate, and Nordic
+    UART)::
+
+        HR_SERVICE = bluetooth.UUID(0x180D)
+        HR_CHAR = (bluetooth.UUID(0x2A37), bluetooth.FLAG_READ | bluetooth.FLAG_NOTIFY,)
+        UART_SERVICE = bluetooth.UUID('6E400001-B5A3-F393-E0A9-E50E24DCCA9E')
+        UART_TX = (bluetooth.UUID('6E400003-B5A3-F393-E0A9-E50E24DCCA9E'), bluetooth.FLAG_READ | bluetooth.FLAG_NOTIFY,)
+        UART_RX = (bluetooth.UUID('6E400002-B5A3-F393-E0A9-E50E24DCCA9E'), bluetooth.FLAG_WRITE,)
+        (
+         (hr,),
+         (tx, rx,),
+        ) = bt.gatts_register_services(
+                                       (
+                                        (HR_SERVICE, (HR_CHAR,),),
+                                        (UART_SERVICE, (UART_TX, UART_RX,),),
+                                       )
+                                      )
+
+    The three value handles (``hr``, ``tx``, ``rx``) can be used with
+    ``gatts_read``, ``gatts_write``, and ``gatts_notify``.
+
+    *Note:* Advertising must be stopped before registering services.
+
+.. method:: BLE.gatts_read(value_handle)
+
+    Reads the local value for this handle (which has either been written by
+    ``gatts_write`` or by a remote central).
+
+.. method:: BLE.gatts_write(value_handle, data)
+
+    Writes the local value for this handle, which can be read by a central.
+
+.. method:: BLE.gatts_notify(conn_handle, value_handle, [data])
+
+    Notifies a connected central that this value has changed and that it should
+    issue a read of the current value from this peripheral.
+
+    If ``data`` is specified, then the that value is sent to the central as part
+    of the notification, avoiding the need for a separate read request.
```

Comment:
Maybe add:

    Note that *data* can be different to the data that is stored in the local value.  The local value will not be updated.